File: framework/scala/src/main/scala/soda/scala/web/setup/ClassLoaderCache.scala
package soda.scala.web.setup

import soda.scala.web.Logger

import scala.collection.mutable

object ClassLoaderCache {

  private val loaderMap = mutable.Map[String, ClassLoader]()

  def remove(path: String): Unit = this.synchronized {
    loaderMap.remove(path)
  }

  def setupForJar(key: String, jarBytes: Array[Byte]): Unit = this.synchronized {
    val parent = Thread.currentThread().getContextClassLoader
    val loader = new JarFileBytesClassLoader(jarBytes, parent)
    loaderMap(key) = loader
  }

  def getForJar(key: String): ClassLoader = this.synchronized {
    loaderMap(key)
  }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/setup/ClassLoaderCache.kt
package soda.kotlin.web.setup

object ClassLoaderCache {

    private val loaderMap = mutableMapOf<String, ClassLoader>()

    @Synchronized
    fun remove(path: String) {
        loaderMap.remove(path)
    }

    @Synchronized
    fun setupForJar(key: String, jarBytes: ByteArray) {
        val parent = Thread.currentThread().contextClassLoader
        loaderMap[key] = JarFileBytesClassLoader(jarBytes, parent)
    }

    @Synchronized
    fun getForJar(key: String): ClassLoader? {
        return loaderMap[key]
    }

}